src/graphforge/executor/evaluator.py

```
@@ -226,8 +228,46 @@ def _evaluate_string_function(func_name: str, _args: list[CypherValue]) -> Cyphe
+        # SUBSTRING(string, start) or SUBSTRING(string, start, length)
+        if not isinstance(args[0], CypherString):
+            raise TypeError(f"SUBSTRING expects string, got {type(args[0]).__name__}")
+        if not isinstance(args[1], CypherInt):
+            raise TypeError("SUBSTRING start must be integer")
+
+        string = args[0].value
+        start = args[1].value
+
+        if len(args) == 3:
+            if not isinstance(args[2], CypherInt):
+                raise TypeError("SUBSTRING length must be integer")
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)